syntax = "proto3";

package cuda_learning;

option go_package = "github.com/jrb/cuda-learning/proto/gen";

// W3C Trace Context for distributed tracing
message TraceContext {
  string traceparent = 1 [json_name = "traceparent"];
  string tracestate = 3 [json_name = "tracestate"];
}

// Enum for filter types
enum FilterType {
  FILTER_TYPE_UNSPECIFIED = 0;
  FILTER_TYPE_NONE = 1;
  FILTER_TYPE_GRAYSCALE = 2;
  // Future: FILTER_TYPE_BLUR = 5, FILTER_TYPE_EDGE_DETECT = 10, etc.
}

// Enum for accelerator types
enum AcceleratorType {
  ACCELERATOR_TYPE_UNSPECIFIED = 0;
  ACCELERATOR_TYPE_GPU = 1;
  ACCELERATOR_TYPE_CPU = 2;
}

// Enum for grayscale conversion algorithms
enum GrayscaleType {
  GRAYSCALE_TYPE_UNSPECIFIED = 0;
  GRAYSCALE_TYPE_BT601 = 1;      // ITU-R BT.601 (SDTV): Y = 0.299R + 0.587G + 0.114B
  GRAYSCALE_TYPE_BT709 = 2;      // ITU-R BT.709 (HDTV): Y = 0.2126R + 0.7152G + 0.0722B
  GRAYSCALE_TYPE_AVERAGE = 3;    // Simple average: Y = (R + G + B) / 3
  GRAYSCALE_TYPE_LIGHTNESS = 4;  // Lightness: Y = (max(R,G,B) + min(R,G,B)) / 2
  GRAYSCALE_TYPE_LUMINOSITY = 5; // Luminosity: Y = 0.21R + 0.72G + 0.07B (similar to BT709)
}

